### PR TITLE
Fix #34 handle 66 Unicode non-characters and surrogates correctly

### LINKED ISSUE #34: utf8proc does not correctly handle the 66 Unicode "noncharacters"
`utf8proc` considers the 66 Unicode "noncharacters" to not be valid, however the Unicode standard specifically says that they are valid code points, and need to be handled correctly in conforming software.

### PR DESCRIPTION
**Pull request by ScottPJones**
This change also improves the performance of the functions utf8proc_iterate and utf8proc_codepoint_valid, and fixes surrogate handling in utf8proc_encode_char.

### CODE REVIEW

**Comment by pao** [comment]
`make check` failed on Travis, possible encoding problem? (Fancy that in a Unicode library.)

**Comment by pao** [comment]
Or possibly it's now testing non-characters because the definition changed.

**Comment by ScottPJones** [comment]
The errors are something to do with graphemes, which I didn't think my changes would even affect.
There are also a number of warnings about a wcwidth function, on both gcc and clang, that I didn't touch :sad:

**Comment by nalimilan** [comment]
Given that both Travis builds failed (AppVeyor does not run the tests), it doesn't look like a random failure. It doesn't look unlikely to me that modifying an essential function like `utf8proc_encode_char` could change the behaviour of tests. Even if you didn't change these functions, they probably call the modified code indirectly.

**Comment by ScottPJones** [comment]
@nalimilan What makes me wonder a bit, is all the warning messages in the travis compilation...
I'll dig into it though

**Comment by stevengj** [comment]
Would also be good to add a regression test to the `tests/` directory.

**Comment by ScottPJones** [comment]
Note: I figured out why the grapheme test failed... but it doesn't make me very happy about utf8proc.
I managed to work around it, but basically, the tests *depended* on the utf8 encoding being broken!
(it encoded some values as 0xFF and 0xFE, which it then used as markers, but left that bad encoding
for *all* callers to the utf8proc_encode_char function).
@stevengj This already has fairly complete tests..., are you talking about additional tests in the julia tests unit tests?

**Comment by stevengj** [comment]
Yes, the `CHARBOUND` option was documented as inserting `0xFF` bytes as delimiters.    I'd be happy to deprecate this functionality, since I find `utf8proc_grapheme_break` to be much more useful (because it doesn't require making a copy of the string etcetera just to detect grapheme breaks).   

Alternatively, we could change it to insert NUL bytes as delimiters, but it would be good to change the API (e.g. to `UTF8PROC_GRAPHEME`) in that case to force a breaking change for code relying on this feature.

By test case, I mean creating an additional test program that checks for regressions on handling of these non-characters and surrogates.

File: test/graphemetest.c [hunk 1]
```diff
@@ -49,7 +51,7 @@ int main(int argc, char **argv)
                 else
                     i++;
             }
-            glen = utf8proc_map(utf8, j, &g, UTF8PROC_CHARBOUND);
+            glen = utf8proc_map(utf8, j, &g, UTF8PROC_CHARBOUND | UTF8PROC_OLDENCODE);
```
Comment (stevengj) [inline test/graphemetest.c]: Shouldn't `UTF8PROC_CHARBOUND` imply `UTF8PROC_OLDENCODE`?  Why have a separate flag?

Comment (ScottPJones) [inline test/graphemetest.c]: I'd missed that (about CHARBOUND...).
OK, I'll try to get some testing in... (in my copious spare time! :grinning:)
Would it be possible to have extra testing as a follow-up PR?

Comment (stevengj) [inline test/graphemetest.c]: In general, it's good practice to merge patches only given regression tests and (where needed) documentation.

Comment (ScottPJones) [inline test/graphemetest.c]: Yes, true... I'll add some further testing, and make sure it gets all of the edge cases that were broken before... there are no API changes though.

File: utf8proc.c [hunk 2]
```diff
@@ -186,6 +186,37 @@ UTF8PROC_DLLEXPORT utf8proc_ssize_t utf8proc_encode_char(utf8proc_int32_t uc, ut
 }
 
 /* internal "unsafe" version that does not check whether uc is in range */
```
Comment (stevengj) [inline utf8proc.c]: Any non-`static` functions should be prefixed with `utf8proc_` in order to prevent namespace pollution.

File: utf8proc.c [hunk 3]
```diff
@@ -165,12 +168,9 @@ UTF8PROC_DLLEXPORT utf8proc_ssize_t utf8proc_encode_char(utf8proc_int32_t uc, ut
...
     return 2;
-  } else if (uc == 0xFFFF) {
-    dst[0] = 0xFF;
-    return 1;
-  } else if (uc == 0xFFFE) {
-    dst[0] = 0xFE;
-    return 1;
+  } else if (((utf8proc_uint32_t)uc)-0xd800 < 0x800) {
+    // Surrogates not valid code point, can't be encoded!
+    return 0;
```
Comment (stevengj) [inline utf8proc.c]: I wonder if it makes more sense to be permissive about this...

Comment (ScottPJones) [inline utf8proc.c]: I don't think so, those are totally bad conversions...

Comment (ScottPJones) [inline utf8proc.c]: Or, did you mean for surrogates, not the 0xffff ones I removed?

Comment (stevengj) [inline utf8proc.c]: I meant for surrogates, which can appear from incorrect conversions of UTF-16 to UTF-8 (or Java-like "modified UTF-8"); I'm wondering what the tradeoffs here are for being permissive in the encoding step, so that we least provide the option of allowing such nonsense.

(We definitely want to detect them and throw an error *somewhere*, e.g. if `UTF8PROC_REJECTNA` is requested, but in that case it is optional.)

Comment (ScottPJones) [inline utf8proc.c]: OK, my take on this, is that for *validity* checking (unless we have some optional flags to make it relaxed... default strict), it should always be strict.  I think only conversion functions should potentially be relaxed, and then only for the two semi-standard variants, Modified UTF-8 (i.e. like Java), and CESU-8, like
Oracle, MySQL, and others (and that the conversion functions in Base only *produce* 100% correct Unicode... if somebody wants to produce something like Modified UTF-8 or CESU-8, they can easily write that in a package...)

File: utf8proc.c (hunk 2 above)
Comment (ScottPJones) [inline utf8proc.c]: OK, that came about because I first did the changes based on what I found in my julia directory... only later did I see that the code really lived in JuliaLang/utf8proc instead of JuliaLang/julia, and missed fixing those... Sorry!

File: utf8proc.c (hunk 3 above)
Comment (stevengj) [inline utf8proc.c]: Except this isn't a validity checking step per se (I agree that the utf8proc_map function should check for surrogates), this is an encoding function.  The question is whether to barf if the user specifically requests to encode a surrogate char.

I just want to be careful about changing the behavior here.  This is a behavior change in an existing API, and whether it is a bug fix is a matter of perspective.

Comment (ScottPJones) [inline utf8proc.c]: OK, I see your point... (it doesn't matter to me as much anymore, because I now would like to just cut Julia's dependence on utf8proc.c anyway... make it pure Julia code if possible, with maybe just a C data file... [but not the one from utf8proc_data.c]).   I'll change the surrogate part back.

File: utf8proc.c (hunk 2 above)
Comment (ScottPJones) [inline utf8proc.c]: Not sure I understand your question about test functions (may be total lack of sleep and jet lag...).  Could you explain further?  Thanks!

File: test/tests.h [hunk 4]
```diff
@@ -45,9 +46,9 @@ size_t encode(char *dest, const char *buf)
                dest[d] = 0; /* NUL-terminate destination string */
                return i + 1;
           }
-          check(sscanf(buf + i, "%x", &c) == 1, "invalid hex input %s", buf+i);
+          check(sscanf(buf + i, "%x", (unsigned int *)&c) == 1, "invalid hex input %s", buf+i);
           i = j; /* skip to char after hex input */
-          d += utf8proc_encode_char(c, (utf8proc_uint8_t *) (dest + d));
-     } while (1);
+          d += encode_char(c, (utf8proc_uint8_t *) (dest + d));
```
Comment (tkelman) [inline test/tests.h]: need to put utf8proc_ prefix back

Comment (ScottPJones) [inline test/tests.h]: It's actually, not putting `utf8proc_` prefix back, but rather adding `utf8proc_unsafe_`.
The version that starts with just `utf8proc_` would not work correctly...  My jetlagged brain missed that test case :sad:

File: utf8proc.c (hunk 2 above)
Comment (stevengj) [inline utf8proc.c]: `unsafe_encode_char` is used in `tests.h`

Comment (ScottPJones) [inline utf8proc.c]: Ah, that was because of that graphemetest, that used the broken version to map 0xffff -> 0xff, and 0xfffe -> 0xfe.

Comment (stevengj) [inline utf8proc.c]: The `tests.h` file is only used for parsing the input test files from Unicode consortium into UTF-8 strings.  The `GraphemeBreakTest.txt` file from the Unicode consortium contains (unpaired) surrogate "code points", presumably for testing UTF-16 code.   The test program encodes these into (invalid) UTF-8, and then `utf8proc_map` (correctly) signals a `UTF8PROC_ERROR_INVALIDUTF8` error.

The graheme test doesn't use 0xFFFF or 0xFFFE, as far as I can tell.  I'm not sure why these cases were included in the `encode` function.  (Maybe some kind of BOM thing?  Not that BOM is needed or recommended for UTF-8 data.)

**Comment by ScottPJones** [comment]
@stevengj Could you also please take a look at my new PR #39 here?
I added some reasonably extensive testing of both the `codepoint_valid` and `iterate` functions...
I did this as a separate PR, so you could easily see the issues I found with the current utf8proc.c code...
Thanks greatly!

File: utf8proc.c [hunk 5]
```diff
@@ -186,7 +188,38 @@ UTF8PROC_DLLEXPORT utf8proc_ssize_t utf8proc_encode_char(utf8proc_int32_t uc, ut
...
+      dst[0] = 0xF0 + (uc >> 18);
+      dst[1] = 0x80 + ((uc >> 12) & 0x3F);
+      dst[2] = 0x80 + ((uc >> 6) & 0x3F);
+      dst[3] = 0x80 + (uc & 0x3F);
+      return 4;
+   } else return 0;
+}
+
+/* internal "unsafe" version that does not check whether uc is in range */
+static const utf8proc_property_t *utf8proc_unsafe_get_property(utf8proc_int32_t uc) {
```
Comment (stevengj) [inline utf8proc.c]: Since this is a `static` function, there is no need to put the `utf8proc_` prefix.   It's good to preserve the convention that exported (non-`static`) functions get the prefix, whereas non-exported functions have no prefix.

**Comment by stevengj** [comment]
Once 39 is merged, can you update the tests to also test for noncharacter validity and include the updated tests in this PR?

**Comment by ScottPJones** [comment]
#39 already does test for noncharacter validity...  If you run it now, it will show you all the problems you have if you *don't* merge this one :grinning:

File: utf8proc.c (hunk 5 above)
Comment (ScottPJones) [inline utf8proc.c]: I added those because I mistakenly thought that's what you were asking me to do... I'll go back and  put those back... (this is what you get for programming on 3 days of only 2 hours of sleep!)

**Comment by stevengj** [comment]
@ScottPJones, then it should be merged into this PR (and squashed/rebased).

**Comment by stevengj** [comment]
Oh, you should also update the `.gitignore` file with the names of the new test programs (so that `git status` will ignore them).

**Comment by ScottPJones** [comment]
But then, isn't it harder to show issues that haven't been fixed yet?
Actually, that's something that I think is a problem with the way the current unit testing works...
I think that you should be able to independently of fixes, be able to add unit tests for any problems that have been found... and the unit test framework should check for regressions on any particular test, while saving and displaying information about those that fail, instead of terminating the testing process...
(I worked this way for many years, it was *very* effective in not losing track of issues in the product, while showing where progress was being made in cleaning up old issues)

**Comment by ScottPJones** [comment]
Oh, didn't know about .gitignore... Thanks!

**Comment by stevengj** [comment]
Committing tests that fail is not really compatible with the Travis/Github workflow.  If you don't merge the PRs, then we are forced to either (a) merge 39 first, in which case you break the build (tests passing/failing is binary on Github) or (b) merge 35 first, in which case we are merging before testing.

**Comment by timholy** [comment]
> I think that you should be able to independently of fixes, be able to add unit tests for any problems that have been found.

I agree this would be a major improvement. There have been quite a few times where I've had to omit a test that should have been there simply because some other bug made it fail. It would be really nice to have a "todo list" of failing tests.

That said, certainly our current framework doesn't support this.

**Comment by ScottPJones** [comment]
@stevengj The way the unit tests would work within the Travis/GitHub framework is fairly simple... A "failure" for Travis would only be when there is a *regression* on a test that passed previously...
New tests show what the behavior is supposed to be, and the ones that don't match, get counted up,
and so you'd see something like:
```
Unicode Tests:
     Line 30, 'isvalid(UTF8String, b"\xff") -> false' failed, returned true
     .... 
  1000 passed, 50 failed, 3.53 seconds
```

What do you think?

**Comment by stevengj** [comment]
@ScottPJones, that's reasonable, but since you have to build this facility "manually" I wouldn't bother on a project as small as utf8proc.  For Julia, it might be worthwhile to extend `Base.Test` with a `@todo` test (for something that is expected to fail) or similar.

**Comment by ScottPJones** [comment]
@stevengj Thanks for all the review!   Hope I got the move of #39 here done correctly...

File: test/tests.h [hunk 6]
```diff
@@ -33,10 +33,11 @@ size_t skipspaces(const char *buf, size_t i)
    separated by whitespace, and terminated by any character not in
    [0-9a-fA-F] or whitespace, then stores the corresponding utf8 string
    in dest, returning the number of bytes read from buf */
+utf8proc_ssize_t unsafe_encode_char(utf8proc_int32_t uc, utf8proc_uint8_t *dst);
 size_t encode(char *dest, const char *buf)
 {
      size_t i = 0, j, d = 0;
-     do {
+     for (;;) {
```
Comment (stevengj) [inline test/tests.h]: BTW, is there some substantive reason you changed this?  Don't most compilers generate exactly the same code for `for (;;) {...}` and `do {...} while(1)`?

Usually, it's best to avoid purely stylistic changes in patches (unless you have a patch that is purely for a global style change); on the contrary, one strives for a patch that matches the surrounding code style.

**Comment by ScottPJones** [comment]
@stevengj  You're quite right, but maybe this should be raised as an issue?  To improve unit testing for Julia?
We actually had a system that kept track of the results of previous unit tests, in CSV files... so it wasn't really "manual", but 1) I didn't build the unit testing framework, 2) it's not the sort of programming I really want to do :grinning:, but I know it can be done and is very very useful, and would be good if *somebody* smart could implement it... (sort of intermediate beginner project? or Summer of Code?)

**Comment by tkelman** [comment]
There are lots of nice small usable test frameworks out there that are better than what we have now - we could just pick one and start using it. I like the one that libgit2 uses for example: https://github.com/vmg/clar

File: test/tests.h (hunk 6 above)
Comment (ScottPJones) [inline test/tests.h]: I don't even remember doing that... that sort of thing happens on autopilot... I date myself :grinning:, but when I started, not all C compilers optimized `while(1)` away!

Comment (ScottPJones) [inline test/tests.h]: Anyway, do I need to change that back and resquash?

**Comment by ScottPJones** [comment]
@tkelman  I'm pretty ignorant of a lot of what's out there in OSS land... (too many years with "not invented here" mentality where I worked... unfortunately I wasn't able to change that...)
Does vmg/clar have the capabilities that I described, of checking for regressions on each individual test?

File: utf8proc.c [hunk 7]
```diff
@@ -165,12 +171,8 @@ UTF8PROC_DLLEXPORT utf8proc_ssize_t utf8proc_encode_char(utf8proc_int32_t uc, ut
...
     dst[1] = 0x80 + (uc & 0x3F);
     return 2;
-  } else if (uc == 0xFFFF) {
-    dst[0] = 0xFF;
-    return 1;
-  } else if (uc == 0xFFFE) {
-    dst[0] = 0xFE;
-    return 1;
+  // Note: we allow encoding 0xd800-0xdfff here, so as not to change
+  // the API, however, these are actually invalid in UTF-8
```
Comment (stevengj) [inline utf8proc.c]: In that case, do we still need `utf8proc_unsafe_encode_char`?

**Comment by tkelman** [comment]
> checking for regressions on each individual test?

Not sure exactly what you mean, but the main thing a nicer test framework like clar (and there are even fancier ones out there, I'm sure) would do is run all the tests and give you a summary of failures at the end.

**Comment by ScottPJones** [comment]
I just started looking at clar, but (in my really quick 50,000 feet flying look), it didn't seem to have built in tracking of previous results... but 1) I'm jetlagged, 2) need to look further...
The way the InterSystems test suite worked (IIRC) is that it would tag a potential release point (for the nightly build, for example), using Perforce, and run the tests, saving the results of all the tests...
(it also had files that had the correct results).  It would then compare (and it had to check if tests were added or deleted) each item, to see if the result changed...  if it changed, and didn't match the *correct* result, that was deemed a regression, and that build would be considered failed.
You'd get a nice report with how many new tests passed, how many new tests failed (along with more information to help debug each individual one), how many old tests passed, and how many failed (but did not change the result), and how many changed behavior and failed...  you could also run the test suite locally... or subsections of it...

**Comment by ScottPJones** [comment]
Oh, I forgot to say, it compared with the last known good build along that branch... then if there are no regressions detected, the build is marked OK, and it's results (if different), are saved and become the new point for future comparisons in the branch.

**Comment by ScottPJones** [comment]
Thanks!
